src/generic/datavgen.cpp

```diff
@@ -3192,6 +3356,303 @@ int wxDataViewMainWindow::GetLineHeight( unsigned int row ) const
 }
 
 
+void RowRange::Add(const unsigned int idx)
+{
+    RangeEntity *rng = NULL;
+
+    size_t count = m_ranges.GetCount();
+    size_t rngIdx = 0;
+    while (rngIdx < count)
+    {
+        rng = &m_ranges[rngIdx];
+
+        if (idx >= rng->from && rng->to >= idx)
+        {
+            // index already in range
+            return;
+        }
+
+        if (rng->from == idx + 1)
+        {
+            rng->from = idx;
+            CleanUp(rngIdx);
+            return;
+        }
+        if (rng->to == idx - 1)
+        {
+            rng->to = idx;
+            CleanUp(rngIdx);
+            return;
+        }
+
+        if (rng->from > idx + 1)
+        {
+            break;
+        }
+
+        rngIdx += 1;
+    }
+//    wxLogMessage("New Range: %d" , count);
+
+    RangeEntity newRange;
+    newRange.from = idx;
+    newRange.to = idx;
+    m_ranges.Insert(newRange, rngIdx);
+}
+
+void RowRange::RemoveFrom(const unsigned int idx)
+{
+    RangeEntity *rng = NULL;
+
+    size_t count = m_ranges.GetCount();
+    size_t rngIdx = 0;
+    while (rngIdx < count)
+    {
+        rng = &m_ranges[rngIdx];
+        if (rng->from >= idx)
+        {
+            m_ranges.RemoveAt(rngIdx);
+            count = m_ranges.GetCount();
+            continue;
+        }
+        if (rng->to >= idx)
+        {
+            rng->to = idx - 1;
+        }
+
+        rngIdx += 1;
+    }
+}
+
+
+void RowRange::CleanUp(int idx)
+{
+    RangeEntity *rng = NULL;
+    RangeEntity *prevRng = NULL;
+
+    size_t count = m_ranges.GetCount();
+    size_t rngIdx = 0;
+    if (idx > 0)
```

Comment:
This check, `idx - 1` below and `idx + 1` in the loop all look like a code smell to me. I'm not sure at all if there are no off-by-1 errors here and it would be really better to define precisely what the index is because it's not clear which row range exactly does it correspond to.